SimpLL: Inverse condition pattern fixes

PLukas2018 opened the pull request:
This PR fixes multiple problems with the inverse condition pattern:
1. The pattern could cause a false negative in a code with a combined condition because it always "skiped" a compare instruction with an inverse condition regardless of what the instruction users were. Example which caused false negative:
    ```c
    // old
    void f() {                                                               
        if (5 == 5 || 10 == 10) {                                            
            return true;                                                     
        }                                                                    
        else return false
    }
    // new
    void f() {                                                               
        if (5 != 5 || 10 != 10) {                                            
            return true;                                                     
        }                                                                    
        else return false
    }
    ```
2. It did not correctly handle inverse conditions with negations, because it searched for the current matching pair in the inverse condition list instead of the previous one when deciding if to add the pair to the list or remove it (more in [bc6abbe](https://github.com/viktormalik/diffkemp/commit/bc6abbe0f702c3b37687d3da8c5a76615dbeaa4e)). Eg it did not handle a case described in [212993d](https://github.com/viktormalik/diffkemp/commit/212993d):
    ```
    %2 = icmp eq %0, %1  | %2 = icmp ne %0, %
                         | %3 = xor %2, true
    br %2 %T %F          | br %3 %T %F
    ```
3. It did not check if an inverse condition pattern is enabled before checking if `not` instruction could not be a case of an inverse condition pattern.

* This PR also adds logging for the inverse condition pattern.

* Note: There is a problem with multiple `not` instructions which it does not handle properly because it would be necessary to find the original cmp instruction in the version of the function which does not have `not` instructions. It could be possible to fix it, but when I tried to compile programs to include the `not` (respectively `xor`) instruction -- I was unable to get such a program because the compiler usually just inverse the condition in the compare instruction `!(a!=b) -> a==b` or just swapped the branches. 

* While experimenting with it, I found out that it also has not been able to handle (with `-O1` optimisation and passes which are used for build-kernel) cases when the condition is saved to a variable, eg something like this:
    ```c
    // old
    bool a(int b, int c, int d) {
	int e=b<0;	
	if (e) {
		return false;
	}
	else {
		return true;
	}
    }
    // new
    bool a(int b, int c, int d) {
	int e=!(b<0);	
	if (e) {
		return true;
	}
	else {
		return false;
	}
    }
    ```
    The problem are `store` and `load` instructions.

PLukas2018 commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -940,26 +946,46 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
+                //                      | %3 = not %2 (%3 = xor %2, true)
+                //                      | %4 = not %3 (%4 = xor %3, true)
+                // br %1, l %T, l %F    | br %4, l %F, l %T
                 sn_mapL.erase(&*InstL);
                 sn_mapR.erase(&*InstR);
```
Also I am not sure, if this should not erase only the `not` instruction or maybe even the original `cmp` instruction?

PLukas2018 replied:
Thanks for the explanation, it makes sense. Now I am thinking if it should not also be erased from `mappedValuesBySn`, but I think it is not necessary, because the next instructions will overwrite it .

viktormalik replied:
That's correct but I think that we should erase it anyways - just to be sure that there is nothing left from the (incorrect) comparison of the last 2 instructions. Ideally, we'd use `undoLastInstCompare` but we can't because we don't want to control the `InstL` and `InstR` iterators in a slightly different way.

Good catch btw! Can you please add the `mappedValuesBySn` erase?

PLukas2018 replied:
Thanks, good idea.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -1959,3 +1985,21 @@ bool DifferentialFunctionComparator::collectBinaryOperands(
+        const Instruction *inst) const {
+    for (auto user : inst->users()) {
+        auto userInstruction = dyn_cast<Instruction>(user);
+        if (isa<BranchInst>(userInstruction) || isLogicalNot(userInstruction)) {
+            checkInverseCondUsers(userInstruction);
+        } else {
+            return false;
+        }
```
This doesn't seem correct, you're not using the result of the recursive call. Also no need for the recursive call in case of `BranchInst`.
Perhaps should be like this?
```suggestion
        if (isLogicalNot(userInstruction) && !checkInverseCondUsers(userInstruction)) {
            return false;
        } else if (!isa<BranchInst>(userInstruction)) {
            return false;
        }
```
We could possibly join the conditions together, but it may be less readable.

PLukas2018 replied:
Thanks, I somehow did not realise or overlooked that...
I used your suggestion but slightly modified it (there was a problem when it was `not` instruction).

viktormalik replied:
Good point, I see it now.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -940,26 +946,46 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
                 // list. Otherwise, add them.
-                size_t erased = inverseConditions.erase(matchingPair);
+                size_t erased = inverseConditions.erase(oldMatchingPair);
+                LOG_VERBOSE_EXTRA((erased ? "Removed" : " Added")
+                                  << " inverse condition\n");
+                LOG_INDENT();
+                LOG_VERBOSE_EXTRA("L: " << (*(matchingPair.first)) << "\n");
+                LOG_VERBOSE_EXTRA("R: " << (*(matchingPair.second)) << "\n");
+                LOG_UNINDENT();
```
This visually decouples the two related operations of updating `inverseConditions`. Better move it below the `insert` line?

PLukas2018 replied:
Thanks, I edited it and I think it is better.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -940,26 +946,46 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
                 sn_mapL.erase(&*InstL);
                 sn_mapR.erase(&*InstR);
 
                 std::pair<const Value *, const Value *> matchingPair;
+                std::pair<const Value *, const Value *> oldMatchingPair;
```
Nice catch! I completely missed this one.

I wonder if we could come up with better names.
Suggestion:
```suggestion
                std::pair<const Value *, const Value *> condPair;
                std::pair<const Value *, const Value *> origCondPair;
```
Maybe even a comment explaining the purpose of each of these would be useful.

PLukas2018 replied:
Thanks, I added the comment and used your suggestion, but I used `prevCondPair` instead of `origCondPair`, because if there are 2 or more `not` instructions the old pair will not be the original condition but it would look like something like this `{not ..., nullptr}` because it is not currently able to find the original instruction. 

I am thinking of fixing it but when I tried to compile some programs with negation (`!`), after compilation there were none `not` (`xor`) instructions. Also currently it would not be able to handle if `not` instructions would be on both sides. But I am not sure if it makes sense to fix this.

viktormalik replied:
Yeah, LLVM does a good job in removing successive negations. You could try to use `-O0` to see if it helps but we still apply some transformations in SimpLL which may remove those.

Anyways, since this is so hard to produce in the IR, it practically shouldn't occur in real code and therefore I don't think it's worth fixing.

PLukas2018 replied:
I did not manage to get LLVM IR containing `not` (xor) instruction even with `-O0`, so I will leave it like this.

viktormalik reviewed (changes requested):
These are great bug discoveries and fixes, many thanks!

One note on the commit order: the golden rule should be that all tests always pass after each commit (helps when hunting bugs down the history). IIUC, that's not the case here as the first commit adds a test which is only fixed by one of the later commits. If I'm right, could you please reorder the commits? Thanks!

Other than that, I've just some minor comments.

PLukas2018 commented:
> One note on the commit order: the golden rule should be that all tests always pass after each commit (helps when hunting bugs down the history). IIUC, that's not the case here as the first commit adds a test which is only fixed by one of the later commits. If I'm right, could you please reorder the commits? Thanks!

Thanks, did not know that.

I addressed the comments. Fixed few other things I overlooked:
- forgot to check users in `not` instruction,
- wrong logging in `not` instruction (did log current pair even when the old was removed),
- add an update of number of compared instructions.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -939,29 +946,64 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
-                // If the conditions are already inverse, remove them from the
-                // list. Otherwise, add them.
-                size_t erased = inverseConditions.erase(matchingPair);
+                // If the previous pair was are already inverse,
```
```suggestion
                // If the previous pair already had inverse instructions,
```

PLukas2018 replied:
Thanks.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -939,29 +946,64 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
+
+                LOG_VERBOSE_EXTRA((erased ? "Removed" : " Added")
+                                  << " inverse condition\n");
+                std::pair<const Value *, const Value *> &pairToLog =
+                        (erased) ? prevCondPair : condPair;
```
Would this work?
```suggestion
                auto &pairToLog = (erased) ? prevCondPair : condPair;
```

viktormalik reviewed (commented):
LGTM, just a couple of small nits. Also please add the `undoLastInstCompare` call.

PLukas2018 commented:
Thanks, I fixed the problems.
